client/src/components/WebAuthn/Register.js

```diff
@@ -36,17 +29,29 @@ class Register extends Component {
     this.handleStartRegistration = this.handleStartRegistration.bind(this);
   }
 
-  /**
-   * keyData may not be passed through on initial render, and we can't start without it.
-   *
-   * @todo Re-evaluate whether this code is required.
-   */
-  componentWillReceiveProps() {
-    const { keyData } = this.props;
-    const { view } = this.state;
+/**
+ * React 16 deprecates componentWillReceiveProps replacing it with getDerivedStateFromProps
+ * for setting state in a functional manner. This function is used by both
+ * {@see componentWillReceiveProps} and {@see getDerivedStateFromProps} in this component
+ * in order to provide the same functionality across different (supported) build versions
+ *
+ * We rely on a prop called `keyData` being present, however it may not be set upon the first
+ * render of the component. We can check and wait for it here.
+ *
+ * @param {Object} newProps incoming props for this component
+ * @param {Object} currentState current stated of this component
+ */
+  static getDerivedStateFromProps(newProps, currentState) {
+    if (currentState.view === VIEWS.LOADING && newProps.keyData) {
+      return { view: VIEWS.READY };
+    }
+    return null;
+  }
 
-    if (view === VIEWS.LOADING && keyData) {
-      this.setState({ view: VIEWS.READY });
+  componentWillReceiveProps(nextProps) {
+    const stateChange = Register.getDerivedStateFromProps(nextProps, this.state);
+    if (stateChange) {
+      this.setState(stateChange);
```

Comment:
Sorry, I hadn't read all of these changes before I posted the last comment.

If I'm reading the logic correctly, construction will set `view` to VIEWS.READY if keyData is in the props, otherwise LOADING. The original logic in componentWillReceiveProps() was checking if the `view` is VIEWS.LOADING and there is keyData, then reverting it back to READY again, which is what the construction should did.

When is the keyData prop provided if not during construction?

Are you able to use `componentDidUpdate()` or something like that instead of `componentWillReceiveProps()`?

Reply:
This has caught me out in the last couple of days too.

keyData is supplied AFTER the `fetch` call from the parent `Verfiy` component (from `silverstripe/mfa`).
That is `startVerifcation` or whatever it's called.